lib/ecto/repo/schema.ex

```diff
@@ -96,11 +101,45 @@ defmodule Ecto.Repo.Schema do
 
     if has_query? do
       rows = plan_query_in_rows(rows, header, adapter)
-      {rows, header, placeholder_vals_list}
+      {rows, header, placeholder_vals_list, counter}
     else
-      {rows, header, placeholder_vals_list}
+      {rows, header, placeholder_vals_list, counter}
     end
   end
+  defp extract_header_and_fields(repo, %Ecto.Query{select: %Ecto.Query.SelectExpr{expr: {:%{}, _ctx, args}}} = query, _schema, _dumper, _autogen_id, _placeholder_map, adapter, opts) do
+    header = Enum.map(args, &elem(&1, 0))
+
+    {query, opts} = repo.prepare_query(:insert_all, query, opts)
```

Comment:
Remember to update the docs/typespec of the prepare_query callback in Ecto.Repo!